### DISCUSSION TITLE
run_until_stalled that doesn't slow down top speed?

### DISCUSSION THREAD

**Discussion by FLL-Team-24277** [pybricks/support#1726] [General]
Is there a way to use stall detection with a duty_limit other than 100% that doesn't slow the motor? 

```python
from pybricks.pupdevices import Motor
from pybricks.parameters import (
    Port,
    Direction,
)
from pybricks.tools import wait

leftAttachmentMotor = Motor(Port.B)

 # default is accel=1000, torque = 199
print(leftAttachmentMotor.control.limits())

leftAttachmentMotor.run_time(time=4000, speed=1000) # 166RPM
wait(1000)
leftAttachmentMotor.control.limits(acceleration=10000, torque=50)
leftAttachmentMotor.run_time(time=4000, speed=1000) # 88 RPM
wait(1000)
leftAttachmentMotor.control.limits(acceleration=10000, torque=199)
leftAttachmentMotor.run_until_stalled(speed=1000, duty_limit=100) # 166 RPM
wait(1000)
leftAttachmentMotor.run_until_stalled(speed=1000, duty_limit=50) # 88 RPM
```

**Comment by laurensvalk** [comment]
Yes, but in a slightly different way.

When you think about it, running at the maximum speed requires maximum power (duty cycle). If you limit it to half the power, how would it run at top speed?

But you can still measure the load. So you can `run(1000)` and then wait until the load value exceeds a threshold of your choosing.

If you don't mind the direction of the load, use `abs(motor.load())` to assess the value. It has the same units and approximate maximum values as the torque limit. If you do know the direction (friction is usually opposite to movement), you can make the threshold even more sensitive.

**Reply by FLL-Team-24277** [reply]
Wow, how did I miss that? With a variable name like "duty_cycle" of course I should have seen that it will lower the speed. Thanks!!!

But now I am curious. How does the motor report the load? How does pybricks know at a certain moment in time that the motor is expending 95 n-m of torque? Are you using the electrical current as a proxy?

**Reply by laurensvalk** [reply]
I apologize if I came across as condescending, I did not mean to :smile: I'm aware that the duty cycle limit could use some better explanations in the docs. Maybe `power_limit` would be clearer.

> But now I am curious. How does the motor report the load? How does pybricks know at a certain moment in time that the motor is expending 95 n-m of torque? Are you using the electrical current as a proxy?

The electrical current can unfortunately not be measured per motor in these hubs. When actively controlling the motor (similar to a PID controller), the load is a measure for how hard the controller has to work to keep up with the target speed or angle.

If you print the load in a loop, you can experiment by slowing down the motor as it tries to move, or push it away from its holding target to see the estimated load. When the PID controller is inactive (`coast`, `brake`, or `dc`) it uses a slightly less accurate estimate that uses the physical model of the motor to estimate load.

**Reply by laurensvalk** [reply]
Here is a demo for how sensitive load measurement can be: https://www.youtube.com/watch?v=QmijLZzLjyk

**Reply by FLL-Team-24277** [reply]
100% was not complaining or anything like that! That was my attempt at self-deprecation :) Thank you VERY VERY much for your help with pybricks!!!

I suppose power_limit would be a bit better, but I still feel a little silly for not recognizing "duty_cycle".
